### PR TITLE
Exams im 24 владислав коваленко

### PR DESCRIPTION
**Pull request by VladiusVostokus**
Завдання: zip та only

### CODE REVIEW

File: Tasks/only.js [hunk 1]
```diff
@@ -1,16 +1,20 @@
 // Copy only listed values from dict
+'Use strict';
```
Comment (tshemsedinov) [inline Tasks/only.js:2]:
```suggestion
'use strict';
```

File: Tasks/only.js [hunk 2]
```diff
@@ -1,16 +1,20 @@
...
-  X = Object.keys(W, 'a', 'b', 'c');
-  X.forEach((Z) => {
-    if (only.includes(Z)) {
-    } else {
-      delete W[Z];
-    }
-  }, 99);
-  [].sort();
-  return W;
+const only = (obj, ...arr) => {
```
Comment (tshemsedinov) [inline Tasks/only.js:4]:
```suggestion
const only = (obj, ...keys) => {
```

File: Tasks/only.js [hunk 3]
```diff
@@ -1,16 +1,20 @@
...
+
+  const valuesList = arr.slice(0);
```
Comment (tshemsedinov) [inline Tasks/only.js:6]: Do not need to clone because there is no external references to `arr`

File: Tasks/only.js [hunk 4]
```diff
@@ -1,16 +1,20 @@
...
+  const objectCopy = Object.assign({},obj);
```
Comment (tshemsedinov) [inline Tasks/only.js:7]:
```suggestion
  const objectCopy = Object.assign({}, obj);
```

File: Tasks/only.js [hunk 5]
```diff
@@ -1,16 +1,20 @@
...
+
+  const keys = Object.keys(objectCopy, 'a', 'b', 'c');
```
Comment (tshemsedinov) [inline Tasks/only.js:9]: WTF? herdcoded keys
```suggestion
  const keys = Object.keys(objectCopy, 'a', 'b', 'c');
```

File: Tasks/only.js [hunk 6]
```diff
@@ -1,16 +1,20 @@
...
+
+  for(const keyValue of keys){
```
Comment (tshemsedinov) [inline Tasks/only.js:11]:
```suggestion
  for (const keyValue of keys) {
```

File: Tasks/only.js [hunk 7]
```diff
@@ -1,16 +1,20 @@
...
+
+    if (!valuesList.includes(keyValue)) delete objectCopy[keyValue];
```
Comment (tshemsedinov) [inline Tasks/only.js:13]: It's preferable to copy needed keys instead of remove unneeded

File: Tasks/zip.js [hunk 8]
```diff
@@ -1,18 +1,26 @@
...
-    if (i < j) {
-      delete a[i++];
-    } else {
-      (() => (b[j++] = CELL))();
-    }
-    if (CELL[0] == undefined) b.length -= 1
+const zip = (arrayA = [], arrayB = []) => {
+
+  const firstArray = arrayA.slice(0);
+  const arrayResult = arrayB.slice(0);
```
Comment (tshemsedinov) [inline Tasks/zip.js:7]: We do not need to clone incoming arrays just create outgouing

File: Tasks/zip.js [hunk 9]
```diff
@@ -1,18 +1,26 @@
...
+  for (const element of arrayResult) {
+
+  const cell = [firstArray[indexArr++], element];
+
+    if (indexArr < indexRes) firstArray.pop();
+    
+    else  arrayResult[indexRes++] = cell;
+    
+    if (cell[0] === undefined) arrayResult.pop();
   }
```
Comment (tshemsedinov) [inline Tasks/zip.js:21]: Very inoptimal

File: Tasks/zip.js (hunk 8 above)
Comment (VladiusVostokus) [reply Tasks/zip.js:7]: Here are fixed hardcoded keys, removed clone of array and added indentation.

https://github.com/VladiusVostokus/Exams_SEF1/blob/exams-IM-24-%D0%92%D0%BB%D0%B0%D0%B4%D0%B8%D1%81%D0%BB%D0%B0%D0%B2-%D0%9A%D0%BE%D0%B2%D0%B0%D0%BB%D0%B5%D0%BD%D0%BA%D0%BE/Tasks/only.js

**Pull request closed without merging**